qmail-local.c

```diff
@@ -79,35 +95,54 @@ void maildir_child(dir)
 char *dir;
 {
  unsigned long pid;
- unsigned long time;
- char myhost[64];
+ struct timeval time;
+ char host[64];
  char *s;
  int loop;
  struct stat st;
  int fd;
  substdio ss;
  substdio ssout;
+ stralloc hostsa = {0};
+ size_t hostnamemaxlen = 256;
 
  sig_alarmcatch(sigalrm);
  if (chdir(dir) == -1) { if (error_temp(errno)) _exit(1); _exit(2); }
  pid = getpid();
- myhost[0] = 0;
- gethostname(myhost,sizeof(myhost));
+ if (!stralloc_ready(&hostsa,hostnamemaxlen)) temp_nomem();
+ if (!gethostname(hostsa.s,hostnamemaxlen)) error_temp(errno);
+ hostsa.s[hostnamemaxlen-1] = 0;
+
+ s = hostsa.s;
+ for (loop = 0; loop < str_len(hostsa.s); ++loop)
+  {
+   if (hostsa.s[loop] == '/')
+    {
+     if (!stralloc_cats(&hostname,"\\057")) temp_nomem();
+     continue;
+    }
+   if (hostsa.s[loop] == ':')
+    {
+     if (!stralloc_cats(&hostname,"\\072")) temp_nomem();
+     continue;
+    }
+   if (!stralloc_append(&hostname,s+loop)) temp_nomem();
+  }
+
  for (loop = 0;;++loop)
   {
-   time = now();
+   gettimeofday(&time, NULL);
    s = fntmptph;
    s += fmt_str(s,"tmp/");
-   s += fmt_ulong(s,time); *s++ = '.';
-   s += fmt_ulong(s,pid); *s++ = '.';
-   s += fmt_strn(s,myhost,sizeof(myhost)); *s++ = 0;
+   s += fmt_ulong(s,time.tv_sec); *s++ = '.';
```

Comment:
This adds both '.' and 'M', I don't thinkg that is intentional.

Reply:
The old Maildir format is "time.pid.hostname:flags". By replacing "pid" with more stuff, including a sub-second time measurement, this patch changes it to "time.MμsecPpid.host:flags". So time.tv_sec and time.tv_usec are both needed, but one comes before the first dot and the other comes after it.

Reply:
Please put all of them in different lines to improve readability. If you want to group the sections than insert newlines.

Reply:
Well-behaving mail software would ignore all the pseudo-random part of the name, so I suppose adding letters in addition to `.` and `0-9` is fine.
Also, I think I remember there are already software around using letters in there.